### ISSUE TITLE
Proposal: Better printing of parent objects

### ISSUE THREAD

**Issue by thofma** [oscar-system/Oscar.jl#1540]
# The problem
Currently, we have one method to print parent objects, which is used everywhere. This has a major drawback, since the julia printing system often expects that objects can (alternatively) be printed in one line. This is exploited by the printing system for example when tuples are printed or in a `Vector`. Here is an example:

```julia
julia> A = [1 2; 3 4] # normal printing takes as many lines as it wants
2×2 Matrix{Int64}:
 1  2
 3  4

julia> (A, A) # printing A inside a tuple 
([1 2; 3 4], [1 2; 3 4])

julia> [A, A] # printing A inside a vector
2-element Vector{Matrix{Int64}}:
 [1 2; 3 4]
 [1 2; 3 4]
```

Many of our objects only have the "take as many lines you want" mode, which leads to some awful output. Take for example the following (taken from #1538)

```julia
julia> affine_algebra(RG)
(Quotient of
===========
Multivariate Polynomial Ring in 7 variables y[1], y[2], y[3], y[4], ..., y[7] over K
graded by 
Abelian group with structure: Z
via
y[1] -> [2], y[2] -> [2], y[3] -> [2], y[4] -> [3], ..., y[7] -> [3]

by the ideal
============
Ideal with 9 generators, Map with following data
Domain:
=======
Quotient of
===========
Multivariate Polynomial Ring in 7 variables y[1], y[2], y[3], y[4], ..., y[7] over K
graded by 
Abelian group with structure: Z
via
y[1] -> [2], y[2] -> [2], y[3] -> [2], y[4] -> [3], ..., y[7] -> [3]

by the ideal
============
Ideal with 9 generators
Codomain:
=========
Multivariate Polynomial Ring in x[1], x[2], x[3], x[4] over K
graded by 
Abelian group with structure: Z
via
x[1] -> [1], x[2] -> [1], x[3] -> [1], x[4] -> [1])
```
Notice the awkward `[...] generators, Map [...]` in the middle of nowhere.

# Possible solution

## Compact printing when printing in "one line"

I propose that we print things in one line if the printing system expects them in one line. I would even go a step further in that we print them *supercompactly* in this case. This new mode will be used recursively.

### Example - Polynomial ring

If `R` is a (graded) polynomial ring, then it will have the following supercompact printing
```
(Graded) Polynomial ring in $(nvars(R)) variables over $(supercompact(coefficient_ring(R))
```
So far example, if number fields print supercompact as "[a] number field", then this would be
```
Polynomial ring in 3 variables over a number field
```
(One has to be a bit careful, it seems that "a" must be part of the printing of the ring somehow.)

### Example - Invariant ring
If `R` is a invariant ring, it could be
```
Invariant ring of $(supercompact(group(R)) over $(coefficicent_ring(R))
```

### Example - Map
If `f` is a map, it could be
```
Map
```
or
```
Map from $(supercompact(domain(f)) to $(supercompact(codomain(f))
```

### Example - Quotient ring

If `R` is a `MPolyQuoRing`, it could be
```
Quotient of $(supercompact(base_ring(R))$
```
or
```
Quotient of $(supercompact(base_ring(R))$ by $(supercompact(ideal(R))$
```

## Do not change the default printing

I think printing the object as we do now still has it benefits, so I would not want to change it (merely clean it up a bit). There must be a way to inspect the object. Hence, I would propose that

```julia
julia> R
[...]
```
still prints as it does now and the supercompact mode would be only used when `R` appears in e.g. a tuple,
```julia
julia> (R, R)
($(supercompact(R)), $(supercompact(R)))
```
So,
```julia
julia> Q, p = affine_algebra(RG)
(Quotient of graded polynomial ring in 7 variables over number field [by ideal generated by 9 elements], Map from quotient of graded polynomial ring in 7 variables [by ideal generated by 9 elements] over number field to graded polynomial ring in 4 variables)

julia> Q
[the usual stuff for Q]
```

(The text in `[ ]` is optional, depending on how quotients should be printed supercompactly, see below.)

Alternatively, we could make the supercompact printing the 'default' printing everywhere, e.g. also for the single `Q` above, but I think it would not be intuitive. If I type `Q` and hit enter, I want to see the details of `Q`.

# Comments and open questions

This proposal here touches not upon the question of whether to print variable names like `K` or `G` if these variables are bound to some objects. This is orthogonal.

If we want to go down this route, there are some questions concerning the supercompact printing of the individual types
- Should a quotient ring print its ideal?
- Should an ideal print the corresponding ring?
- Probably many other

On a technical note, this could be implemented relatively easy and in a backwards compatible manner by passing some `:supercompact => true` (or whatever name we decide on) around in the expressify system.

I am also happy it discuss it next week in person. I wanted to put it here before someone sinks more time into improving the printing of selective objects. It should be addressed consistently.

CC: @joschmitt @wdecker @fingolfin @fieker @tthsqe12

**Comment by joschmitt** [comment]
One thing I would like to add is that there should be some "limit" how big the prints of substructures are allowed to be. For example, for a quotient ring `Q = R/I` it can be very annoying if typing `Q` prints 10000 generators of `I`. If I want to get this piece of information I can ask for `modulus(Q)`.
In some cases like this, one could also use the proposed "supercompact printing", but I don't think this would work everywhere. For example, if a quotient ring only prints as `Quotient ring of polynomial ring modulo ideal`, then that's not helpful either.

I guess I'm proposing a 3-tier system: The "standalone" mode printing everything that's necessary to define the object (for example used for an ideal `I` if one just types `I`), the "substructure" mode shortening this information (e.g. leaving away the ambient ring for ideals and printing the generators as `[f, g, ..., h]`), and then the "supercompact" mode for tuples.

And one more thing: commas should be forbidden in anything printed in tuples (supercompact). Right now the number of entries of the tuple `(Multivariate Polynomial Ring in x, y, z over Rational Field, fmpq_mpoly[x, y, z])` could lie between 1 and 4...

**Comment by wdecker** [comment]
Currently we have two choices w.r.t. printing:

...;    no printing at all
...     printing (often horrible)

Would it be possible to introduce a tiered system of printing, from no printing via supercompact via ....   ... to "all" ???

If so, there should be some default way of printing which in my eyes should be rather supercompact than "all".

Here is another horrible example:
```
julia> L = normalization(A)
2-element Vector{Tuple{MPolyQuo{fmpq_mpoly}, Oscar.MPolyAnyMap{MPolyQuo{fmpq_mpoly}, MPolyQuo{fmpq_mpoly}, Nothing, MPolyQuoElem{fmpq_mpoly}}, Tuple{MPolyQuoElem{fmpq_mpoly}, MPolyQuoIdeal{fmpq_mpoly}}}}:
 (Quotient of Multivariate Polynomial Ring in T(1), T(2), T(3), x, y over Rational Field by ideal(2*T(2)*y + T(3)*y + x, T(1)*x - T(2)*y - T(3)*y, -T(1)*y + T(2)*x - y, T(1)*y^2 - T(1)*y - T(3)*x - y, T(2)*y^3 + T(2)*y^2 - T(3)*x^2 + T(3)*y^3, T(1)^2 + T(1), T(1)*T(2), T(1) + T(2)^2 + 1, T(1)*T(3) + T(2) + T(3), -T(1) + T(2)*T(3) - 1, T(1)*y + T(1) + T(3)^2 + 1, -x^4 + x^2*y^3 - x^2*y^2 + y^5), Map with following data
Domain:
=======
A
Codomain:
=========
Quotient of Multivariate Polynomial Ring in T(1), T(2), T(3), x, y over Rational Field by ideal(2*T(2)*y + T(3)*y + x, T(1)*x - T(2)*y - T(3)*y, -T(1)*y + T(2)*x - y, T(1)*y^2 - T(1)*y - T(3)*x - y, T(2)*y^3 + T(2)*y^2 - T(3)*x^2 + T(3)*y^3, T(1)^2 + T(1), T(1)*T(2), T(1) + T(2)^2 + 1, T(1)*T(3) + T(2) + T(3), -T(1) + T(2)*T(3) - 1, T(1)*y + T(1) + T(3)^2 + 1, -x^4 + x^2*y^3 - x^2*y^2 + y^5), (-2*x^3 + x*y^3 - x*y^2, ideal(x^3 + x*y^2, -x^2*y + y^4, x^2*y^2 + x^2*y, -2*x^3 + x*y^3 - x*y^2)))
 (Quotient of Multivariate Polynomial Ring in x, y over Rational Field by ideal(x), Map with following data
Domain:
=======
A
Codomain:
=========
Quotient of Multivariate Polynomial Ring in x, y over Rational Field by ideal(x), (1, ideal(1)))
```

**Comment by thofma** [comment]
It seems that we have varying opinions/expectations for the printing. It might be best to discuss this next week in person.

**Comment by simonbrandhorst** [comment]
+1 for @thofma s proposition. It is certainly an improvement.

**Comment by fingolfin** [comment]
The current plan is that someone (@simonbrandhorst) will prepare (using e.g. HackMD) a collection of example inputs with proposals for how the "ideal" output could look like. We can then use this as basis for discussions to hammer out a proposal we can actually implement.

**Comment by simonbrandhorst** [comment]
#2148

**Comment by simonbrandhorst** [comment]
Closed via #2148
